### PR TITLE
Add 3D terrain with animated directions route 

### PR DESCRIPTION
**Pull request by tobrun**
This PR adds an example of an animated route on 3D terrain:
 - route geometry is fetched from the Mapbox directions API and visualize with a LineLayer
 - Animate a LocationLayer to visualize the end user location
 - Animate a FreeCameraOptions across the route 

ezgif com-video-to-gif(24)

### CODE REVIEW

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        )
+        +locationIndicatorLayer(LOCATION_LAYER_ID) {
+          topImage(FOREGROUND_ICON)
+          bearingImage(BACKGROUND_ICON)
+          imagePitchDisplacement(5.0)
+          topImageSize(1.5)
+          bearingImageSize(1.5)
+        }
+        +image(FOREGROUND_ICON) {
+          bitmap(BitmapFactory.decodeResource(resources, R.drawable.mapbox_mylocation_icon_default))
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:89]: nit: could use the generateBitmap function as well.

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+      .accessToken(getString(R.string.mapbox_access_token))
+      .build()
+    client.enqueueCall(object : Callback<DirectionsResponse> {
+      override fun onResponse(
+        call: Call<DirectionsResponse>,
+        response: Response<DirectionsResponse>
+      ) {
+        animateRoute(
+          LineString.fromPolyline(
+            response.body()!!.routes()[0].geometry()!!,
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:140]: nit: we could avoid using `!!` to handle errors more gracefully.

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+      LOCATION_LAYER_ID
+    )
+
+    // cache the camera
+    val camera = mapboxMap.getFreeCameraOptions()
+
+    // get a reference to elevations of the animation
+    // note: to call getElevation the coordinate needs to be in the viewport
+    val startElevation = mapboxMap.getElevation(POINT_START)!!.toDouble()
+    val endElevation = mapboxMap.getElevation(POINT_END)!!.toDouble()
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:181]: nit: could avoid using `!!` and handle the `null` gracefully by providing fallback values.

**Review by pengdev — COMMENTED** [review]
Great example! LGTM with a few nits.

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    // load satellite style and add terrain with a line layer to visualize the route
+    mapboxMap.loadStyle(
+      style(styleUri = Style.SATELLITE_STREETS) {
+        +rasterDemSource(SOURCE) {
+          url(TERRAIN_URL_TILE_RESOURCE)
+          tileSize(512)
+        }
+        +terrain(SOURCE) {
+          exaggeration(1.7)
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:72]: Do we actually need it?
If yes - move to const.

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+      }
+
+      // use phase that trails behind to be able to create a tilted camera
+      val cameraLocation = TurfMeasurement.along(
+        lineString, routeDistance * cameraPhase,
+        TurfConstants.UNIT_KILOMETERS
+      )
+
+      // calculate the elevation of the user location
+      val elevation = minOf(maxElevation, (maxElevation * phase) + minElevation)
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:217]: It seems that `elevation` will grow linearly from `minElevation` to `maxElevation` for some time phase starting from `phase == 0` and starting from some concrete phase will become `maxElevation` till `phase == 1.0`. 
Could you please clarify logic here?

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+      locationLayer.location(
+        listOf(
+          cameraLookingAt.latitude(),
+          cameraLookingAt.longitude(),
+          elevation
+        )
+      )
+
+      // calculate an elevation difference between the user position and the camera
+      val elevationDifference = BASE_CAMERA_ELEVATION + elevation
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:229]: Is it indeed difference? Looks more like a sum to me 😄

**Review by kiryldz — APPROVED** [review]
Spectacular example!
Asked some questions to better understand the underline logic but in general LGTM

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt (hunk 2 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:140]: I will optimize for it but I disagree that we should always replace `!!` with graceful error handling. In the SDK, we should never use `!!`. In the test app we are allowed to use it. My personal believe is that it's better to fail fast and you at least know something is wrong. Versus silently ignoring and printing out logcat message.

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt (hunk 4 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:72]: need it, no but it looks cooler

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt (hunk 2 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt:140]: for context I'm replacing it with this code:

```
        response.body()?.let { body ->
          body.routes()[0].geometry()?.let {
            animateRoute(
              LineString.fromPolyline(it, PRECISION_6)
            )
            return
          }
        }
        throw RuntimeException("Not able to retrieve a directions route")
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt (hunk 5 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt]: correct

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt (hunk 6 above)
Comment (tobrun) [reply app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt]: updated the comments

**Review by pengdev — APPROVED** [review]

**Pull request merged**